lsbatch/man1/busers.1
.ds ]W %
.ds ]L
.nh
.TH busers 1 "Volclava Version 1.0 - June 2025"
.br
.SH NAME
\fBbusers\fR - displays information about users and user groups
.SH SYNOPSIS
.BR
.PP
.PP
\fBbusers\fR\fB \fR[\fB-w\fR]\fR\fB \fR[\fIuser_name \fR... | \fIuser_group\fR ... | \fBall\fR]\fB
.BR
.PP
\fBbusers\fR\fB \fR[\fB-h\fR | \fB-V\fR]
.SH DESCRIPTION
.BR
.PP
.PP
\fB\fRDisplays information about users and user groups.
.PP
By default, displays information about the user who runs the 
command.
.SH OPTIONS
.BR
.PP
.TP 
\fIuser_name\fR ... |\fB \fR\fIuser_group \fR... |\fB \fR\fBall
\fR
.IP
Displays information about the specified users or user groups, or about 
all users if you specify all.

.TP
\fB-w
\fR
.IP
Prints pending job thresholds for users/usergroups, showing MPEND, PJOBS, and MPJOBS fields.

.TP 
\fB-h
\fR
.IP
Prints command usage to stderr and exits.


.TP 
\fB-V
\fR
.IP
Prints Lava release version to stderr and exits. 


.SH OUTPUT
.BR
.PP
.PP
A listing of the users and user groups is displayed with the following 
fields: 
.SS USER/GROUP 
.BR
.PP
.PP
The name of the user or user group. 
.SS JL/P 
.BR
.PP
.PP
The maximum number of job slots that can be processed 
simultaneously for the specified users on each processor. 
These job slots are used by running and suspended jobs or by pending 
jobs which have jobs slots reserved for them.  This job limit is 
configured per processor so that multiprocessor hosts have more job 
slots. If the dash character (-) is displayed, there is no limit. JL/P is 
defined in the Lava configuration file lsb.users(5). 
.SS MAX 
.BR
.PP
.PP
The maximum number of job slots that can be processed concurrently 
for the specified users' jobs. These job slots are used by running 
and suspended jobs or by pending jobs which have job slots reserved 
for them.  If the character `-' is displayed, there is no limit. MAX 
is defined by the MAX_JOBS parameter in the configuration file 
lsb.users(5). 
.SS NJOBS 
.BR
.PP
.PP
The current number of job slots used by specified users' jobs. A parallel 
job that is pending is counted as \fIn\fR job slots for it will use \fIn\fR job slots 
in the queue when it is dispatched. 
.SS PEND 
.BR
.PP
.PP
The number of pending job slots used by jobs of the specified users.
.SS RUN 
.BR
.PP
.PP
The number of job slots used by running jobs of the specified users. 
.SS SSUSP 
.BR
.PP
.PP
The number of job slots used by the system-suspended jobs of the 
specified users. 
.SS USUSP
.BR
.PP
.PP
The number of job slots used by user-suspended jobs of the specified 
users. 
.SS RSV 
.BR
.PP
.PP
The number of job slots used by pending jobs of the specified users 
which have job slots reserved for them. 
.SS MPEND
.BR
.PP
.PP
The pending job slot threshold for the specified users or user groups,
which is defined by MAX_PEND_SLOTS in the configuration file
lsb.users(5).
.SS PJOBS
.BR
.PP
.PP
The number of users' pending jobs.
.SS MPJOBS
.BR
.PP
.PP
The pending job threshold for the specified users or user groups,
which is defined by MAX_PEND_JOBS in the configuration file
lsb.users(5).
.SH SEE ALSO
.BR
.PP
.PP
bugroup(1), lsb.users(5), lsb.queues(5)


ChangeLog
2025-06-16  Mingze Li <limingze.jiayou@bytedance.com>
            Nan Qi <qinan.cn@bytedance.com>
            Guangbo Shu <shuguangbo@bytedance.com>

	* release volclava 2.0.0.

	* Summary of changes:
	1. support fairshare scheduling policy for users at queue level;
	2. support customize unit by configure LSF_UNIT_FOR_LIMITS in lsf.conf;
	3. support conf MAX_PEND_JOBS and MAX_PEND_SLOTS in lsb.users and lsb.params;
	4. support conf DEFAULT_LIMIT_IGNORE_USER_GROUP and SUB_TRY_INTERVAL in lsb.params;
	5. support resource reserve as per-host;
	6. fix loadstop not effect job scheduling;
	7. fix memory leak in bhosts and putEnv();
	8. fix pim hang when format of pim.info file is not in volclava format;
	9. fix sbd hang due to unnecessary popen;
	10. fix lsload -w does not display full hostname;
	11. fix mbatchd coredump when only suspend jobs remain;
	12. fix mem be reserved repeatedly after bresume ususp/ssusp job;
	13. fix running jobs exceed the number of slots due to shared type resource when slotResourceReserve=y;

2025-01-06  Mingze Li  <limingze.jiayou@bytedance.com>
            Nan Qi <qinan.cn@bytedance.com>

	* release volclava 1.0.1.

	* Summary of changes:
	1. bugfix: revert check for tcl result with function in tcl 8.6;
	2. configure: set default limit from cpu to cpu>=0 in lsf.task;
	3. update year of Copyright from 2021-2024 to 2021-2025;
	4. bsub: enhance parameter check of bsub -R;
	5. make source-script active in multiple volclava on the same host;
	6. make install script support ubuntu OS;

2024-11-11  Mingze Li  <limingze.jiayou@bytedance.com>

	* re-open as volclava 1.0.

	* Summary of changes:
	1. feature support:
        - bjobs -UF;
        - bjobs -o/-json;
        - bsub -pack;
        - bsub -Ep;
	2. bugfix:
	    - MXJ not equal with maxCpus when set "!";
	    - lshosts -l segmentation fault;
	    - sbatchd block by greater than 1000 jobs;
	    - prefix not work in rpm install to costomize directory;
	    - fix job slot limit reached while host is free;
	3. define new project name as volclava in related files.
	4. adapt to Ubuntu 20.04 and Rocky 8.10.

2012-01-23  David Bigagli  <david@openlava.net>

	* Created 2.0 branch.

2011-11-07  David Bigagli  <david@openlava.net>

	* lib.host.c Rewrote the caching code to read the hosts file only
	once and add the host in the cache. Removed the cache timeout
	and simplified the entire implementation to use only two interface
	functions: Gethostbyname_() and Gethostbyaddr_().
	Change the old C types u_int and casta to char *
	into the current linux types whenever required.
	Modified the daemons code to use the new host library
	implementation.

2011-10-30  David Bigagli  <david@openlava.net>

	* mbd.policy.c Developed a little round robin scheduler absed on the number of jobs a user has running.

	* openlava.spec  modified the spec file so that autoconf creates
	openlava configuration files and use the outptu variables to make
	the necessary subsititution in the them. Change the post install
	to just erase the package without saving anything.
	Removed the symbolic link as that is something sites have to
	do as they may want to run more versions together, also
	in now the lsf.conf has the version in the openlava
	fundamental variables clearly indicating which version is in use.

2011-09-16  David Bigagli  <david@openlava.net>

	* README removed obsolete information about the source code directory structure.
	Added the README and COPYING file to the rmp file.

2011-09-14  David Bigagli  <david@openlava.net>

	* lsf.h Added the openlava foundation copyright.

2011-09-09  David Bigagli  <david@openlava.net>

	* openlava.spec Updated the spec file to reflect the changes in teh build system. The ownership of all files changed, no longer root, the entire installation tree is owned by the openlava user. The daemons to run in multi user mode must still be started at root which is guaranteed at boot by init.

2011-08-27  David Bigagli  <david@openlava.net>

	* openlavaconfig.h Removed it from the top directory and merge its content into lsf.h which is the openlava master header file. In this way the repository is more GNU like as well.

2011-08-23  David Bigagli  <david@openlava.net>

	* GNU build system. Switched openlava from home made makefiles to the GNU build system.

2011-07-20  David Bigagli  <david@openlava.net>

	* scripts/Makefile: added the installation
	of scripts that manage the integration with
	mpi and openmpi.

2011-07-17  David Bigagli  <david@openlava.net>

	* lib.table.c: Added getClosestPrime() function to
	always size and resize tables as primies.

	* lsbatch/daemons/mbd.init.c (addHost):
	Set the number of CPUs on batch host to be its MXJ.
	If MXJ is not confgiured than use the number of physical
	CPUs as provided by openlava base.


2011-07-16  David Bigagli  <david@openlava.net>

	* Added lsb.hosts

2011-06-16  David Bigagli  <david@openlava.net>

	* Makefiles changed all variables to run install from LAVA
	to OPENLAVA.

	* config/Makefile added a command to install COPYING and README files
	to the LAVA_ROOT/conf directory

2011-06-07  David Bigagli  <david@openlava.net>

	* lava startup script was rewritten, simplified and introduce
	the possibility to run daemons in single user mode.

	* Added a doc directory with original Platform pdf for lava
	administrator and user.

	* lava.spec Rober Stober added a new lava.spec to buidl rpm.

	* Make.misc removed the reference to the architecture x86_64
	and changed the make install to also install the lava sturtup script.


2011-05-02  David Bigagli  <david@openlava.net>

	* Makefile(s): Removed PREFIX variable from all make files and
	introduced Lava specific installation variables in Make.misc.

	* Makefile(s): Made static link for all Lava binaries.

	* Makefiles(s): Changed makefiles not to build shared libraries
	by default and modified the shared libraries to use standard .so
	suffix instead of .So.

2011-04-24  David Bigagli  <david@openlava.net>


lsbatch/cmd/busers.c
/*
 * Copyright (C) 2021-2025 Bytedance Ltd. and/or its affiliates
 *
 * $Id: busers.c 397 2007-11-26 19:04:00Z mblack $
 * Copyright (C) 2007 Platform Computing Inc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of version 2 of the GNU General Public License as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA
 *
 */

#include <unistd.h>

#include "cmd.h"

#define NL_SETN 8

#define USR_NAME_LENGTH    18
#define USR_JL_P_LENGTH    5
#define USR_NJOBS_LENGTH   6
#define USR_PEND_LENGTH    6
#define USR_RUN_LENGTH     6
#define USR_SSUSP_LENGTH   6
#define USR_USUSP_LENGTH   6
#define USR_RSV_LENGTH     6
#define USR_MAX_LENGTH     6
#define USR_MPJOBS_LENGTH  6
#define USR_PJOBS_LENGTH   6
#define USR_MPEND_LENGTH   6

static void display_users (struct userInfoEnt *, int);
static void sort_users (struct userInfoEnt *, int);

static char fomt[200];
static char wflag = FALSE;

void
usage (char *cmd)
{
    fprintf(stderr, I18N_Usage);
    fprintf(stderr, ": %s [-w] [-h] [-V] [user_name ...] [all]\n", cmd);
    exit(-1);
}

int
main (int argc, char **argv)
{
    int cc, numUsers;
    struct userInfoEnt  *usrInfo;
    char **users=NULL, **userPoint;
    int all = FALSE;
    int rc;

    rc = _i18n_init ( I18N_CAT_MIN );

    if (lsb_init(argv[0]) < 0) {
	lsb_perror("lsb_init");
	exit(-1);
    }

    while ((cc = getopt(argc, argv, "wVh")) != EOF) {
        switch (cc) {
            case 'w':
                wflag = TRUE;
                break;
            case 'V':
                fputs(_LS_VERSION_, stderr);
                exit(0);
            case 'h':
            default:
                usage(argv[0]);
        }
    }
    numUsers = getNames (argc, argv, optind, &users, &all, "user");
    if (!all && !numUsers) {
        numUsers = 1;
        userPoint = NULL;
    } else
        userPoint = users;
    if (!(usrInfo = lsb_userinfo (userPoint, &numUsers))) {
        if (lsberrno == LSBE_BAD_USER && userPoint)
            lsb_perror(users[numUsers]);
        else
            lsb_perror(NULL);
        exit (-1);
    }
    display_users(usrInfo, numUsers);
    _i18n_end ( ls_catd );
    exit(0);

}

static void
display_users (struct userInfoEnt *reply, int numReply)
{
    int i;
    int first = TRUE;
    char maxJobs[MAX_CHARLEN], procJobLimit[MAX_CHARLEN];
    char maxPendJobs[MAX_CHARLEN], maxPendSlots[MAX_CHARLEN];

    if (first) {
        first = FALSE;

        if (lsbMode_ & LSB_MODE_BATCH) {
            prtWord(USR_NAME_LENGTH, I18N_USER_GROUP, 0);
            prtWord(USR_JL_P_LENGTH, I18N_JL_P, -1);
            prtWord(USR_MAX_LENGTH,  I18N_MAX, -1);
        }
        else
            prtWord(USR_NAME_LENGTH, I18N_USER, 0);

        prtWord(USR_NJOBS_LENGTH, I18N_NJOBS, -1);
        prtWord(USR_PEND_LENGTH,  I18N_PEND,  -1);
        prtWord(USR_RUN_LENGTH,   I18N_RUN,   -1);
        prtWord(USR_SSUSP_LENGTH, I18N_SSUSP, -1);
        prtWord(USR_USUSP_LENGTH, I18N_USUSP, -1);
        prtWord(USR_RSV_LENGTH,   I18N_RSV,   -1);
        if (wflag) {
            prtWord(USR_MPEND_LENGTH,I18N_MPEND,-1);
            prtWord(USR_PJOBS_LENGTH,I18N_PJOBS,-1);
            prtWord(USR_MPJOBS_LENGTH,I18N_MPJOBS,-1);
        }
        printf("\n");
    }
    if (numReply > 1)
        sort_users (reply, numReply);

    for (i = 0; i < numReply; i++) {

        if (reply[i].procJobLimit < INFINIT_FLOAT) {
            sprintf(fomt, "%%%d.1f ", USR_JL_P_LENGTH);
            sprintf (procJobLimit, fomt, reply[i].procJobLimit);
        }
        else
             strcpy(procJobLimit, prtDash(USR_JL_P_LENGTH) );


        if (reply[i].maxJobs < INFINIT_INT)
            strcpy(maxJobs, prtValue(USR_MAX_LENGTH, reply[i].maxJobs) );
        else
             strcpy(maxJobs, prtDash(USR_MAX_LENGTH) );

        if (reply[i].maxPendJobs < INFINIT_INT)
            strcpy(maxPendJobs, prtValue(USR_MAX_LENGTH, reply[i].maxPendJobs) );
        else
            strcpy(maxPendJobs, prtDash(USR_MAX_LENGTH) );

        if (reply[i].maxPendSlots < INFINIT_INT)
            strcpy(maxPendSlots, prtValue(USR_MAX_LENGTH, reply[i].maxPendSlots) );
        else
            strcpy(maxPendSlots, prtDash(USR_MAX_LENGTH) );

        prtWordL(USR_NAME_LENGTH, reply[i].user);

        if (lsbMode_ & LSB_MODE_BATCH) {
                sprintf(fomt, "%%%ds%%%ds", USR_JL_P_LENGTH,
                                            USR_MAX_LENGTH );
            printf(fomt,
                       procJobLimit, maxJobs);
        };

        if (strcmp(reply[i].user, "default") == 0) {
            printf("%s%s%s%s%s%s",
                   prtDash(USR_NJOBS_LENGTH), prtDash(USR_PEND_LENGTH),
                   prtDash(USR_RUN_LENGTH),   prtDash(USR_SSUSP_LENGTH),
                   prtDash(USR_USUSP_LENGTH), prtDash(USR_RSV_LENGTH)
                   );
            if (wflag) {
                printf("%s%s%s",
                       maxPendSlots, prtDash(USR_PJOBS_LENGTH), maxPendJobs
                );
            }
            printf("\n");
        }
        else if (reply[i].numJobs == -INFINIT_INT) {
            printf("%s%s%s%s%s%s",
                   prtDash(USR_NJOBS_LENGTH), prtDash(USR_PEND_LENGTH),
                   prtDash(USR_RUN_LENGTH),   prtDash(USR_SSUSP_LENGTH),
                   prtDash(USR_USUSP_LENGTH), prtDash(USR_RSV_LENGTH)
                   );
            if (wflag) {
                printf("%s%s%s",
                       maxPendSlots, prtDash(USR_PJOBS_LENGTH), maxPendJobs
                );
            }
            printf("\n");
        }
        else {
            if (wflag) {
                sprintf(fomt, "%%%dd %%%dd %%%dd %%%dd %%%dd %%%dd %%%ds %%%dd %%%ds\n",
                        USR_NJOBS_LENGTH,
                        USR_PEND_LENGTH,
                        USR_RUN_LENGTH,
                        USR_SSUSP_LENGTH,
                        USR_USUSP_LENGTH,
                        USR_RSV_LENGTH,
                        USR_MPEND_LENGTH,
                        USR_PJOBS_LENGTH,
                        USR_MPJOBS_LENGTH);
                printf(fomt,
                       reply[i].numJobs, reply[i].numPEND, reply[i].numRUN,
                       reply[i].numSSUSP, reply[i].numUSUSP, reply[i].numRESERVE,
                       maxPendSlots, reply[i].numPENDJobs, maxPendJobs);
            } else {
                sprintf(fomt, "%%%dd %%%dd %%%dd %%%dd %%%dd %%%dd\n",
                        USR_NJOBS_LENGTH,
                        USR_PEND_LENGTH,
                        USR_RUN_LENGTH,
                        USR_SSUSP_LENGTH,
                        USR_USUSP_LENGTH,
                        USR_RSV_LENGTH );
                printf(fomt,
                       reply[i].numJobs, reply[i].numPEND, reply[i].numRUN,
                       reply[i].numSSUSP, reply[i].numUSUSP, reply[i].numRESERVE);
            }
        }
    }
}

static void
sort_users (struct userInfoEnt *users, int numUsers)
{
    int i, j, k;
    struct userInfoEnt temUser;

    for (k = numUsers / 2; k > 0; k /= 2) {
        for (i = k; i < numUsers; i++) {
            for (j = i-k; j >= 0; j -= k) {
                 if (strcmp(users[j].user, users[j+k].user) < 0)
                     break;
                 memcpy(&temUser, users+j, sizeof (struct userInfoEnt));
                 memcpy(users+j, users+j+k, sizeof (struct userInfoEnt));
                 memcpy(users+j+k, &temUser, sizeof (struct userInfoEnt));
            }
        }
    }
    return;

}

